### PR TITLE
Allow group by on array type

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Relates to https://github.com/crate/crate/issues/11796

Can be supported by lifting restrictions on array type, done in the first commit and passes tests.

However, there are other things to be considered, quote from https://github.com/crate/crate/pull/12582#issuecomment-1145716745:

1) > - probably won't work when using docvalues for optimized group iterators 
2) > -  should be discussed if we want to do that specially when having in mind that we may remove source storage in future. 

on 1 - we use source lookup for arrays - https://github.com/crate/crate/commit/e3d09834721cd2b74022d5b7185fca17aabb0b27, so group by array will not fall to doc values aggregation  - but it works with regular GroupingCollector, so nothing bad here, rather performance thing?

on 2 - In case we drop source - we will update LuceneReferenceResolver for arrays and will update ArrayMapper to use DocValues as a follow up?

3) Was not sure what to do with `GroupByOnArrayUnsupportedException` - I didn't remove it as it's used in all ColumnsRefs, for example:
https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/main/java/io/crate/expression/reference/doc/lucene/LongColumnReference.java#L51
and Optimised iterator https://github.com/crate/crate/blob/ffba331c0f012b8143ad233fe2b036938730b889/server/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java#L324
it seems to me that it's unreachable code(?) and if it's showed it indicates a bug ( for example https://github.com/crate/crate/issues/10045) - but I decided to keep the exception with slightly updated message to understand better a failing scenario if it happens.

### CODE REVIEW

File: server/src/test/java/io/crate/integrationtests/GroupByAggregateTest.java [hunk 1]
```diff
@@ -1417,4 +1417,24 @@ public void test_group_by_on_ignored_column() {
...
+    public void test_group_by_array_type() {
+        execute("create table tbl (arr array(int))");
+        execute("insert into tbl(arr) values ([1, 2, null]), ([2, 1]), ([1, 2]), ([1, 2])");
+        execute("insert into tbl(arr) values ([]::int[]), ([]::int[]), ([]::int[])");
+        execute("insert into tbl(arr) values ([null]), ([null]), ([null]), ([null])");
+        execute("insert into tbl(arr) values (null), (null), (null), (null), (null)");
+        refresh();
+        execute("select arr, count(*) cnt from tbl group by arr order by cnt, arr[1]");
+        assertThat(TestingHelpers.printedTable(response.rows()),
+            Is.is("[1, 2, null]| 1\n" +
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/integrationtests/GroupByAggregateTest.java:1431]: aligned with PG 14.

File: server/src/main/java/io/crate/exceptions/GroupByOnArrayUnsupportedException.java [hunk 2]
```diff
@@ -32,7 +32,7 @@
 public class GroupByOnArrayUnsupportedException extends ElasticsearchException implements UnscopedException {
 
     public GroupByOnArrayUnsupportedException(String columnName) {
-        super("Column \"" + columnName + "\" has a value that is an array. Group by doesn't work on Arrays");
+        super("Column \"" + columnName + "\" has a value that is an array. Group by doc values doesn't work on Arrays");
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/exceptions/GroupByOnArrayUnsupportedException.java:35]: Not sure, code should not enter this section - (docvalues are not supported for arrays and we throw it from there) - just added a small hint where to look for if we see it

File: server/src/test/java/io/crate/expression/reference/doc/IpColumnReferenceTest.java [hunk 3]
```diff
@@ -118,6 +121,12 @@ public void testIpExpression() throws Exception {
     @Test
     public void testIpExpressionOnArrayThrowsException() throws Exception {
         IpColumnReference columnReference = new IpColumnReference(IP_ARRAY_COLUMN);
+//        // in real code we explicitly use
```
[the remark this answers is not in the document]
Comment (matriv) [reply server/src/test/java/io/crate/expression/reference/doc/IpColumnReferenceTest.java:121]: I don't get this comment and code, could you please provide more info?

[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/expression/reference/doc/IpColumnReferenceTest.java:121]: My main concern is whether we can completely get rid of `GroupByOnArrayUnsupportedException` - looks like in the tests we write multiple IP-s to doc values https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/test/java/io/crate/expression/reference/doc/IpColumnReferenceTest.java#L78-L83 to provoke the exception - but if I understand correctly, in real code we don't index arrays into doc values, at least I didn't find anything related to doc values in the `ArrayMapper`. I was looking for smth similar to https://github.com/crate/crate/blob/d5ecbfe0c2a315fd1234b58d8e868c987579413a/server/src/main/java/org/elasticsearch/index/mapper/IpFieldMapper.java#L179 but looks like for arrays we don't use them.
ReferenceResolver also [indicates](https://github.com/crate/crate/pull/4633/files#diff-c8c672d9579dc6acf8097d70b659b28ba858c0dfcffc6590f8c07ace520f3449R107-R109) that we don't expect docvalues for arrays - so we can safely get rid of `GroupByOnArrayUnsupportedException`. Let's also see what @mfussenegger says.

**Review by matriv — APPROVED** [review]
Thank you!

File: server/src/test/java/io/crate/expression/reference/doc/IpColumnReferenceTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (matriv) [reply server/src/test/java/io/crate/expression/reference/doc/IpColumnReferenceTest.java]: @BaurzhanSakhariev Thanks a lot for explaining!

**Pull request merged**